### PR TITLE
AD: Initial test infrastructure

### PR DESCRIPTION
**Pull request by pan-**

### CODE REVIEW

File: doc/adr/0002-define-initial-testing-infrastructure.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+We will use `google test` to unit test services proposed. The tests will be compiled 
+with `Clang` and will generate a code coverage report when run.
+For pull request validation, the code and test suite will be compiled and run by `Circle CI`. 
+Depending on results and code coverage, the submission will be rejected. 
+
+## Consequences
+
+The base of the test infrastructure should be created, it must be based on google test, 
+be compiled with clang run in Circle CI. 
+Maintainers should not merge PR with less than 80% code coverage.
```
Comment (paul-szczepanek-arm) [inline doc/adr/0002-define-initial-testing-infrastructure.md:62]: who are those maintainers you speak of?

**Review by paul-szczepanek-arm — APPROVED** [review]

**Comment by paul-szczepanek-arm** [comment]
IMHO this is a bit too heavy for what this is meant to be. Will this attract contributors or appear as a barrier to them?

File: doc/adr/0002-define-initial-testing-infrastructure.md (hunk 1 above)
Comment (pan-) [reply doc/adr/0002-define-initial-testing-infrastructure.md:62]: Us for now, selected members of the community latter.

**Comment by pan-** [comment]
@paul-szczepanek-arm Thanks for the feedback Paul, that's a proposition and certainly the point that requires the most debate. I put 80% as an arbitrary measure but it can be dropped to lower value or even put has a recommendation. Maybe it should not even be part of this ADR which would just focus on the initial test infrastructure (gtest + mock + circle CI).

**Comment by AGlass0fMilk** [comment]
> IMHO this is a bit too heavy for what this is meant to be. Will this attract contributors or appear as a barrier to them?

The comprehensive testing and CI that many of Mbed's official repositories have is one of their most comforting aspects. My previous bad experiences with the old, untested Mbed BLE services is a motivating factor behind why this repository was created.

The services and extensions here should be high quality, and that quality must be ensured through rigorous testing. I think once the example is set it won't be too hard to leverage existing tests when creating new ones.

That being said, I don't have any idea what level of work is required to attain 80% code coverage :sweat_smile: 

So perhaps we make that a suggested value here and determine a concrete minimum later on once we see what is realistic?

**Comment by pan-** [comment]
@paul-szczepanek-arm @AGlass0fMilk I updated the PR based on your feedback and removed code coverage requirement, this can be the subject of another AD when we have a better understanding of what would be a reasonable threshold and the amount of work required to get there.

**Comment by AGlass0fMilk** [comment]
LGTM

**Review by AGlass0fMilk — APPROVED** [review]

**Pull request merged**